src/persist-types/src/arrow.rs

```diff
@@ -420,6 +420,49 @@ pub struct ArrayIdx<'a> {
     pub array: &'a ArrayOrd,
 }
 
+#[inline]
+fn list_range<'a>(
+    offsets: &OffsetBuffer<i32>,
+    values: &'a ArrayOrd,
+    idx: usize,
+) -> impl Iterator<Item = ArrayIdx<'a>> {
+    let offsets = offsets.inner();
+    let from = offsets[idx].as_usize();
+    let to = offsets[idx + 1].as_usize();
+    (from..to).map(|i| values.at(i))
+}
+
+impl<'a> ArrayIdx<'a> {
+    /// Returns the rough amount of space required for this entry in bytes.
+    /// (Not counting nulls, dictionary encoding, or other space optimizations.)
+    pub fn data_bytes(&self) -> usize {
+        match self.array {
+            ArrayOrd::Null(_) => 0,
+            ArrayOrd::Bool(_) => 1,
+            ArrayOrd::Int8(_) => 1,
+            ArrayOrd::Int16(_) => 2,
+            ArrayOrd::Int32(_) => 4,
+            ArrayOrd::Int64(_) => 8,
+            ArrayOrd::UInt8(_) => 1,
```

Comment:
Personally, I like doing something like `size_of::<u8>()` for stuff like this. Not because I expect the hardcoded values to rot, so much as I find the semantics to be more direct. Ymmv.